File: app/src/main/java/com/example/bilancio/SettingsFragment.java
package com.example.bilancio;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.appcompat.app.AppCompatDelegate;

public class SettingsFragment extends PreferenceFragmentCompat {

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.root_preferences, rootKey);

        // Trova la preferenza del tema utilizzando la sua chiave
        ListPreference themePreference = findPreference("theme");
        if (themePreference != null) {
            themePreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue) {
                    String themeOption = (String) newValue;
                    updateTheme(themeOption);
                    return true;
                }
            });
        }

        // Trova e gestisci la preferenza "Condividi"
        Preference sharePreference = findPreference("share");
        if (sharePreference != null) {
            sharePreference.setOnPreferenceClickListener(preference -> {
                shareApp();
                return true;
            });
        }

        // Trova e gestisci la preferenza "Recensisci questa app"
        Preference reviewPreference = findPreference("review");
        if (reviewPreference != null) {
            reviewPreference.setOnPreferenceClickListener(preference -> {
                reviewApp();
                return true;
            });
        }
    }

    private void updateTheme(String themeOption) {
        switch (themeOption) {
            case "light":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                break;
            case "dark":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                break;
            default:
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);
                break;
        }
    }

    private void shareApp() {
        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        String shareBody = "Dai un'occhiata a Bilancio, un'app fantastica per aiutarti a risolvere gli esercizi di economia!";
        // Aggiungi qui il link diretto al Google Play Store se disponibile
        String shareSub = "Prova Bilancio!";
        shareIntent.putExtra(Intent.EXTRA_SUBJECT, shareSub);
        shareIntent.putExtra(Intent.EXTRA_TEXT, shareBody);
        startActivity(Intent.createChooser(shareIntent, "Condividi con"));
    }

    private void reviewApp() {
        // Sostituisci "com.example.bilancio" con l'ID effettivo del pacchetto della tua app
        Uri uri = Uri.parse("market://details?id=com.example.bilancio");
        Intent goToMarket = new Intent(Intent.ACTION_VIEW, uri);
        // Per provare ad aprire il Google Play Store prima e, se non disponibile, aprire il browser
        goToMarket.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY |
                Intent.FLAG_ACTIVITY_NEW_DOCUMENT |
                Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
        try {
            startActivity(goToMarket);
        } catch (Exception e) {
            startActivity(new Intent(Intent.ACTION_VIEW,
                    Uri.parse("http://play.google.com/store/apps/details?id=com.example.bilancio")));
        }
    }
}

File: app/src/main/java/com/example/bilancio/SettingsFragment.kt
package com.example.bilancio

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.appcompat.app.AppCompatDelegate

class SettingsFragment : PreferenceFragmentCompat() {

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.root_preferences, rootKey)


        val themePreference: ListPreference? = findPreference("theme")
        themePreference?.setOnPreferenceChangeListener { _, newValue ->
            val themeOption = newValue as String
            updateTheme(themeOption)
            true
        }


        val sharePreference: Preference? = findPreference("share")
        sharePreference?.setOnPreferenceClickListener {
            shareApp()
            true
        }


        val reviewPreference: Preference? = findPreference("review")
        reviewPreference?.setOnPreferenceClickListener {
            reviewApp()
            true
        }
    }

    private fun updateTheme(themeOption: String) {
        when (themeOption) {
            "light" -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            "dark" -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            else -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
        }
    }

    private fun shareApp() {
        val shareIntent = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            val shareBody = "Dai un'occhiata a Bilancio, un'app fantastica per aiutarti a risolvere gli esercizi di economia!"
            // Aggiungi qui il link diretto al Google Play Store se disponibile
            val shareSub = "Prova Bilancio!"
            putExtra(Intent.EXTRA_SUBJECT, shareSub)
            putExtra(Intent.EXTRA_TEXT, shareBody)
        }
        startActivity(Intent.createChooser(shareIntent, "Condividi con"))
    }

    private fun reviewApp() {

        val uri = Uri.parse("market://details?id=com.example.bilancio")
        val goToMarket = Intent(Intent.ACTION_VIEW, uri).apply {
            addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY or
                    Intent.FLAG_ACTIVITY_NEW_DOCUMENT or
                    Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
        }
        try {
            startActivity(goToMarket)
        } catch (e: Exception) {
            startActivity(Intent(Intent.ACTION_VIEW,
                Uri.parse("http://play.google.com/store/apps/details?id=com.example.bilancio")))
        }
    }
}


